### PR TITLE
Remove POSIX timer usage

### PR DESCRIPTION
**Pull request by anosach-luxoft**
Roll back previous timer implementation based on `pthread`s because of issues with native POSIX timer. Also, restored implementation had been refactored and simplified.

@AGritsevich, @AGaliuzov, @Kozoriz @Kozoriz, @OHerasym, @pestOO, @malirod, @LuxoftAKutsan, @AByzhynar please review

### CODE REVIEW

File: src/components/utils/src/timer.cc [hunk 1]
```diff
@@ -29,190 +29,129 @@
...
 void timer::Timer::Stop() {
   LOG4CXX_AUTO_TRACE(logger_);
-  sync_primitives::AutoLock auto_lock(lock_);
-  repeatable_ = false;
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK(stop_result);
+  DCHECK(thread_);
+  delegate_.set_stop_flag(true);
+  if (!pthread_equal(pthread_self(), thread_->thread_handle())) {
```
Comment (malirod) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, please avoid direct pthread usage. This code is not portable.

File: src/components/utils/src/timer.cc [hunk 2]
```diff
@@ -29,190 +29,129 @@
...
+  set_stop_flag(false);
+  while (!stop_flag()) {
+    const Milliseconds curr_timeout = timeout();
+    LOG4CXX_DEBUG(logger_, "Milliseconds left to wait: " << curr_timeout);
+    if (sync_primitives::ConditionalVariable::kTimeout == termination_condition_.WaitFor(auto_lock, curr_timeout)) {
+      LOG4CXX_DEBUG(logger_, "Timer has finished counting. Timeout (ms): " << curr_timeout);
+      DCHECK(timer_);
+      timer_->OnTimeout();
+    } else {
+      LOG4CXX_DEBUG(logger_, "Timeout reset force (ms): " << curr_timeout);
```
Comment (malirod) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, on my opinion it's better to save timestamp before wait and after wait log the actual waited time. Currently it's not clear how long the timer actually has spent.

File: src/components/application_manager/src/request_controller.cc [hunk 3]
```diff
@@ -496,12 +495,8 @@ void RequestController::UpdateTimer() {
                    << date_time::DateTime::getmSecs(current_time - end_time)
                    << " Request timeout (sec): "
                    << front->timeout_msec()/date_time::DateTime::MILLISECONDS_IN_SECOND);
-      timer_.Start(0u, false);
+      timer_.Start(0u, true);
     }
-  } else {
-    LOG4CXX_DEBUG(logger_, "Sleep for default sleep time "
-                  << default_sleep_time_ << " milliseconds.");
-    timer_.Start(default_sleep_time_, false);
```
Comment (malirod) [inline src/components/application_manager/src/request_controller.cc]: @anosach-luxoft, please describe in the commit message why the starts are redundant.

File: src/components/utils/src/timer.cc [hunk 4]
```diff
@@ -29,190 +29,129 @@
...
-    const bool stop_result = StopUnsafe();
-    DCHECK_OR_RETURN_VOID(stop_result);
-  }
-  StartUnsafe();
+  Stop();
+  delegate_.set_timeout(timeout);
+  delegate_.set_stop_flag(false);
+  single_shot_ = single_shot;
+  DCHECK(thread_);
+  thread_->start();
```
Comment (malirod) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, will this method wait until the delegate is actually started?

File: src/components/utils/src/timer.cc (hunk 2 above)
Comment (anosach-luxoft) [inline src/components/utils/src/timer.cc]: @malirod it's redundant. We check condvar termination status: `kTimeout`. I think it's enough, don't you?

File: src/components/utils/src/timer.cc (hunk 4 above)
Comment (anosach-luxoft) [inline src/components/utils/src/timer.cc]: @malirod yes, it will

**Comment by Kozoriz** [comment]
@anosach-luxoft , i don`t sure about correct behavior, can you clarify behavior with @pestOO? 
What timer must do when it running now, and we call Start again?

**Comment by anosach-luxoft** [comment]
@AKozoriz, @pestOO is a participant of this PR. In case we call `Start()` on running timer - timer will be restarted (stopped and started again with new timeout)

File: src/components/utils/src/timer.cc (hunk 2 above)
Comment (malirod) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, seems you didn't understand me. To make thing simpler, just remove curr_timeout

File: src/components/utils/src/timer.cc [hunk 5]
```diff
@@ -29,190 +29,129 @@
...
+  DCHECK(thread_);
+  DeleteThread(thread_);
   sync_primitives::AutoLock auto_lock(task_lock_);
   DCHECK(task_);
   delete task_;
+  LOG4CXX_DEBUG(logger_, "Timer " << name_ << " has been destroyed");
 }
 
-void timer::Timer::Start(const Milliseconds timeout, const bool repeatable) {
+void timer::Timer::Start(const Milliseconds timeout, const bool single_shot) {
```
Comment (AGaliuzov) [inline src/components/utils/src/timer.cc]: nice naming `single_shot`:)

File: src/components/utils/src/timer.cc (hunk 4 above)
Comment (AGritsevich) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, you are right, `thread_->start()` really wait until thread started, but in this condition it's not every time signalling when thread started. So, create extra sync object it's really redundant and better find root cause of this problem.

File: src/components/utils/include/utils/timer.h [hunk 6]
```diff
@@ -29,105 +29,116 @@
  * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
  * POSSIBILITY OF SUCH DAMAGE.
  */
-
 #ifndef SRC_COMPONENTS_UTILS_INCLUDE_UTILS_TIMER_H_
 #define SRC_COMPONENTS_UTILS_INCLUDE_UTILS_TIMER_H_
 
-#include <stdint.h>
```
Comment (AByzhynar) [inline src/components/utils/include/utils/timer.h]: @anosach-luxoft Why did you delete this header?

Comment (anosach-luxoft) [inline src/components/utils/include/utils/timer.h]: @AByzhynar please remind me what for do we need it?

**Comment by AGritsevich** [comment]
@pestOO, need your opinion here!
1. Have to be here feature to restart one-shot timer with new time-out, like it was released in old timer?
2. Possibly now change Interfaces `Timer `class?

File: src/components/utils/src/timer.cc (hunk 4 above)
Comment (AGritsevich) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, And I'm not sure that good solution use own copy `start flag` if we have already flag in `ThreadDelegate`.

File: src/components/utils/include/utils/timer.h (hunk 6 above)
Comment (pestOO) [inline src/components/utils/include/utils/timer.h]: @anosach-luxoft, may be for `uint32_t`?

File: src/components/utils/include/utils/timer.h [hunk 7]
```diff
@@ -29,105 +29,116 @@
...
+    explicit TimerDelegate(const Timer* timer);
 
-  /**
-   * @brief method for setting correct timeout.
-   * @param timeout - if it`s value = 0, timeout will be setted to 1
-   * There would be no way to stop thread if timeout in lopper will be 0
-   * and if we puts to timer_create zero timeout then we get sys error(22)
-   */
-  void SetTimeoutUnsafe(const Milliseconds timeout);
+    void set_timeout(const Milliseconds timeout);
```
Comment (pestOO) [inline src/components/utils/include/utils/timer.h]: MAybe also provide doxygen?
For example, Is this time includes `OntImeout` logics routine?
What will be in case of 0 or maximum uint32 value?

File: src/components/utils/include/utils/timer.h [hunk 8]
```diff
@@ -44,20 +44,61 @@ namespace timer {
...
+   */
   class TimerDelegate : public threads::ThreadDelegate {
    public:
+    /**
+     * @brief Constructor
+     * @param timer Timer instance pointer for callback calling
+     */
     explicit TimerDelegate(const Timer* timer);
 
     void set_timeout(const Milliseconds timeout);
```
Comment (pestOO) [inline src/components/utils/include/utils/timer.h]: @anosach-luxoft, 
Maybe also provide doxygen?
For example, Is this time includes `OntImeout` logics routine?
What will be in case of 0 or maximum uint32 value? 
Or seuggetion to follow Timer tinout value

File: src/components/utils/include/utils/timer.h [hunk 9]
```diff
@@ -29,105 +29,116 @@
...
-   * @brief alone function which sends to posix_timer as callee
-   * @param signal_value - structure with parameters of posix_timer callee
+   * @brief Callback called on timeout
    */
-  friend void HandlePosixTimer(sigval signal_value);
+  void OnTimeout() const;
+
+  const std::string name_;
+  TimerTask* task_;
+  mutable sync_primitives::Lock task_lock_;
```
Comment (pestOO) [inline src/components/utils/include/utils/timer.h]: Amy be use `sync_primitives::Atomic<T>`?

File: src/components/utils/src/timer.cc [hunk 10]
```diff
@@ -29,190 +29,129 @@
...
 }
 
 timer::Timer::~Timer() {
   LOG4CXX_AUTO_TRACE(logger_);
-  LOG4CXX_DEBUG(logger_, "Timer is to be destroyed " << name_);
   Stop();
+  DCHECK(thread_);
+  DeleteThread(thread_);
   sync_primitives::AutoLock auto_lock(task_lock_);
   DCHECK(task_);
```
Comment (pestOO) [inline src/components/utils/src/timer.cc]: DCHECK is redundant. a) It not lead to any logic problems) b) deletion null is legal

File: src/components/utils/src/timer.cc [hunk 11]
```diff
@@ -29,190 +29,129 @@
...
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK_OR_RETURN_VOID(stop_result);
-  }
-  StartUnsafe();
+  Stop();
+  delegate_.set_timeout(timeout);
+  delegate_.set_stop_flag(false);
+  single_shot_ = single_shot;
+  DCHECK(thread_);
```
Comment (pestOO) [inline src/components/utils/src/timer.cc]: May be `DCHECK_OR_RETURN`?

File: src/components/utils/src/timer.cc (hunk 4 above)
Comment (pestOO) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, @AGritsevich, @malirod, in case of first start it will really waits for thread creation, and in case of second iteration it looks like it only notify ConditionalVariable

Comment (pestOO) [inline src/components/utils/src/timer.cc]: Hm... But i see that in this case 2 - we generate redundant options for already started thread. Need to investigate this.
https://github.com/smartdevicelink/sdl_core/blob/6916a4ea20083315cbf732f1de3c57916f83213b/src/components/utils/src/threads/posix_thread.cc#L182
+
https://github.com/smartdevicelink/sdl_core/blob/6916a4ea20083315cbf732f1de3c57916f83213b/src/components/utils/src/threads/posix_thread.cc#L214

File: src/components/utils/src/timer.cc [hunk 12]
```diff
@@ -29,190 +29,129 @@
...
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK_OR_RETURN_VOID(stop_result);
-  }
-  if (repeatable_) {
-    StartUnsafe();
-  }
+  delegate_.set_stop_flag(single_shot_);
+  sync_primitives::AutoLock auto_lock(task_lock_);
+  DCHECK(task_);
```
Comment (pestOO) [inline src/components/utils/src/timer.cc]: Maybe `OR_RETURN`?

File: src/components/utils/src/timer.cc [hunk 13]
```diff
@@ -29,190 +29,129 @@
...
   }
-  return false;
+  set_timeout(0);
 }
 
-timer::Milliseconds timer::Timer::GetTimeout() const {
-  sync_primitives::AutoLock auto_lock(lock_);
-  return timeout_ms_;
+void timer::Timer::TimerDelegate::exitThreadMain() {
+  sync_primitives::AutoLock auto_lock(state_lock_);
```
Comment (pestOO) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, lokks liek it will be deadlock - because `threadMain` locks this for whole cycle and the main idea of `exitThreadMain` to stop this `main`

File: src/components/utils/src/timer.cc [hunk 14]
```diff
@@ -29,190 +29,129 @@
...
-  if (StopPosixTimer(timer_)) {
-    is_running_ = false;
-    return true;
+void timer::Timer::TimerDelegate::threadMain() {
+  sync_primitives::AutoLock auto_lock(state_lock_);
+  set_stop_flag(false);
+  while (!stop_flag()) {
+    const Milliseconds curr_timeout = timeout();
+    LOG4CXX_DEBUG(logger_, "Milliseconds left to wait: " << curr_timeout);
+    if (sync_primitives::ConditionalVariable::kTimeout == termination_condition_.WaitFor(auto_lock, curr_timeout)) {
```
Comment (pestOO) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, May be encapsulate this line to `WaitForInteraption` or something like that method?
may be just `WaitFor` with return result `interapted` and `expired`

File: src/components/utils/src/timer.cc (hunk 2 above)
Comment (pestOO) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, agree with @malirod that will be more appropriate to log time elapsed because of interaction instead of waiting timeout.

File: src/components/utils/src/timer.cc [hunk 15]
```diff
@@ -29,190 +29,129 @@
...
+      LOG4CXX_DEBUG(logger_, "Timer has finished counting. Timeout (ms): " << curr_timeout);
+      DCHECK(timer_);
+      timer_->OnTimeout();
+    } else {
+      LOG4CXX_DEBUG(logger_, "Timeout reset force (ms): " << curr_timeout);
+    }
   }
-  return false;
+  set_timeout(0);
```
Comment (pestOO) [inline src/components/utils/src/timer.cc]: @anosach-luxoft,
Why we need to set a 0 value, which will be set to 1u?

Alos it's mena that after timer exit `get_timout` will return 1u instead of set by user value.

File: src/components/utils/src/timer.cc [hunk 16]
```diff
@@ -29,190 +29,129 @@
...
 }
 
-timer::Milliseconds timer::Timer::GetTimeout() const {
-  sync_primitives::AutoLock auto_lock(lock_);
-  return timeout_ms_;
+void timer::Timer::TimerDelegate::exitThreadMain() {
+  sync_primitives::AutoLock auto_lock(state_lock_);
+  termination_condition_.NotifyOne();
```
Comment (pestOO) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, do we need to lock CondVeriable for notifying?

**Comment by pestOO** [comment]
Reviewed

File: src/components/utils/include/utils/timer.h [hunk 17]
```diff
@@ -29,105 +29,72 @@
...
+#include "utils/macro.h"
 #include "utils/lock.h"
 #include "utils/timer_task.h"
+#include "utils/threads/thread.h"
+#include "utils/threads/thread_delegate.h"
 
 namespace timer {
+
 typedef uint32_t Milliseconds;
-/**
```
Comment (LuxoftAKutsan) [inline src/components/utils/include/utils/timer.h]: @anosach-luxoft   Why did you removed comments?

File: src/components/utils/include/utils/timer.h [hunk 18]
```diff
@@ -29,105 +29,72 @@
...
+
+    void threadMain() OVERRIDE;
+    void exitThreadMain() OVERRIDE;
+
+   private:
+    const Timer* timer_;
+    Milliseconds timeout_;
+
+    bool stop_flag_;
+    sync_primitives::Lock state_lock_;
```
Comment (LuxoftAKutsan) [inline src/components/utils/include/utils/timer.h]: @anosach-luxoft  Please write comments with descriptions what this locks locks

File: src/components/utils/include/utils/timer.h [hunk 19]
```diff
@@ -29,105 +29,72 @@
...
-   */
-  void StartUnsafe();
-  bool StopUnsafe();
-
-  /**
-   * @brief alone function which sends to posix_timer as callee
-   * @param signal_value - structure with parameters of posix_timer callee
-   */
-  friend void HandlePosixTimer(sigval signal_value);
+  TimerTask* task_;
```
Comment (LuxoftAKutsan) [inline src/components/utils/include/utils/timer.h]: @anosach-luxoft  If timer is responsible to manage task_ lifecycle? if Timer, it is good idea to use autopointer there.

File: src/components/utils/include/utils/timer.h [hunk 20]
```diff
@@ -29,105 +29,72 @@
...
+  mutable sync_primitives::Lock task_lock_;
+
+  mutable TimerDelegate delegate_;
+  threads::Thread* thread_;
+
+  bool single_shot_;
```
Comment (LuxoftAKutsan) [inline src/components/utils/include/utils/timer.h]: @anosach-luxoft  Atomic bool would be nice

File: src/components/utils/src/timer.cc (hunk 5 above)
Comment (LuxoftAKutsan) [inline src/components/utils/src/timer.cc]: @anosach-luxoft  I am not sure, but in my opinion it would be nice to extract methods to Interface :
Start
StartRepeatable
It is much more clear that some boolean flags.

File: src/components/utils/src/timer.cc (hunk 13 above)
Comment (anosach-luxoft) [inline src/components/utils/src/timer.cc]: @pestOO we call `WaitFor()` in `threadMain()` and it will release lock during waiting. Am I right?

File: src/components/utils/src/timer.cc (hunk 1 above)
Comment (LuxoftAKutsan) [inline src/components/utils/src/timer.cc]: @anosach-luxoft  If timer try to stop itself, it just won't do anything in your realisation. Is it ok?

File: src/components/policy/src/policy/usage_statistics/src/counter.cc [hunk 21]
```diff
@@ -103,7 +103,7 @@ AppStopwatchImpl::AppStopwatchImpl(
 
 void AppStopwatchImpl::Start(AppStopwatchId stopwatch_type) {
   stopwatch_type_ = stopwatch_type;
-  timer_.Start(time_out_ * date_time::DateTime::MILLISECONDS_IN_SECOND, true);
+  timer_.Start(time_out_ * date_time::DateTime::MILLISECONDS_IN_SECOND, false);
```
Comment (LuxoftAKutsan) [inline src/components/policy/src/policy/usage_statistics/src/counter.cc]: @malirod @pestOO 
What do you think about my idea to extend interface with 2 methods:
Start
StartRepeatable 
?

File: src/components/utils/src/timer.cc (hunk 16 above)
Comment (anosach-luxoft) [inline src/components/utils/src/timer.cc]: @pestOO what do you mean `to lock CondVeriable`?? We have to lock `state_lock_` and we do it

File: src/components/utils/src/timer.cc (hunk 14 above)
Comment (anosach-luxoft) [inline src/components/utils/src/timer.cc]: @pestOO, for now `WaitFor()` method returns interruption status (kTimeout, kNoTimeout, etc). Isn't it enough?
You mean we have to add method `WaitForInterruption()`?

File: src/components/utils/include/utils/timer.h (hunk 17 above)
Comment (anosach-luxoft) [inline src/components/utils/include/utils/timer.h]: @LuxoftAKutsan please see 5b988c0

File: src/components/utils/include/utils/timer.h (hunk 20 above)
Comment (anosach-luxoft) [inline src/components/utils/include/utils/timer.h]: @LuxoftAKutsan seems like we don't need to synchronize this variable

File: src/components/utils/include/utils/timer.h (hunk 19 above)
Comment (anosach-luxoft) [inline src/components/utils/include/utils/timer.h]: @LuxoftAKutsan yes, timer is responsible. But our timer class manage this pointer like `auto_ptr` - receives it in constructor and destroys in destructor, so I think `auto_ptr` is redundant here. But I can add if you insist

File: src/components/utils/src/timer.cc (hunk 1 above)
Comment (anosach-luxoft) [inline src/components/utils/src/timer.cc]: @LuxoftAKutsan it will not do anything! It will set stop flag on previous line.

File: src/components/policy/src/policy/usage_statistics/src/counter.cc (hunk 21 above)
Comment (anosach-luxoft) [inline src/components/policy/src/policy/usage_statistics/src/counter.cc]: @LuxoftAKutsan as for me it's redundant, but if you guys insist I'll do it

File: src/components/utils/src/timer.cc (hunk 16 above)
Comment (pestOO) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, I hope Notifing method is thread safety and in this case locking is redundant.

File: src/components/utils/src/timer.cc (hunk 14 above)
Comment (pestOO) [inline src/components/utils/src/timer.cc]: @anosach-luxoft , agree it looks good.

File: src/components/utils/src/timer.cc (hunk 13 above)
Comment (pestOO) [inline src/components/utils/src/timer.cc]: @anosach-luxoft, Sorry, I messed it up. It works according to Posix.

File: src/components/policy/src/policy/usage_statistics/src/counter.cc (hunk 21 above)
Comment (pestOO) [inline src/components/policy/src/policy/usage_statistics/src/counter.cc]: @LuxoftAKutsan, @anosach-luxoft  I agree that it's more OOP way. But I'm not really insist.

**Comment by pestOO** [comment]
@AGritsevich,
> 1. Have to be here feature to restart one-shot timer with new time-out, like it was released in old timer?

Starting timer with new value is looks reasonable.
At least the following case is unappropriated:
```
timer->start(2, SINGLE = TRUE);
// after 2 seconds TimOut occurs
timer->start(4!, SINGLE = TRUE);
// after 21 seconds TimOut occurs
```
Have we test such case?
 
> 2. Possibly now change Interfaces Timerclass?

What do you exactly mean?
> 3. Make Stoptimer every Startcommand?

https://github.com/smartdevicelink/sdl_core/pull/454#discussion_r58364375

File: src/components/policy/src/policy/usage_statistics/src/counter.cc (hunk 21 above)
Comment (malirod) [inline src/components/policy/src/policy/usage_statistics/src/counter.cc]: @LuxoftAKutsan, at the moment I don't see any benefit in from methods instead of one.

Comment (LuxoftAKutsan) [inline src/components/policy/src/policy/usage_statistics/src/counter.cc]: @anosach-luxoft  okey, let it be

File: src/components/utils/include/utils/timer.h (hunk 9 above)
Comment (Kozoriz) [inline src/components/utils/include/utils/timer.h]: @pestOO , who is Amy ? ;)

File: src/components/application_manager/src/request_controller.cc (hunk 3 above)
Comment (Kozoriz) [inline src/components/application_manager/src/request_controller.cc]: @anosach-luxoft , can you reply that message?

File: src/components/utils/include/utils/timer.h (hunk 9 above)
Comment (pestOO) [inline src/components/utils/include/utils/timer.h]: *may be

File: src/components/utils/include/utils/timer.h (hunk 6 above)
Comment (AByzhynar) [inline src/components/utils/include/utils/timer.h]: @anosach-luxoft : @pestOO is absolutely right.

**Comment by anosach-luxoft** [comment]
All, please review added commits since 	5e35c39. PR has been rebased because of conflicts

**Comment by anosach-luxoft** [comment]
Few words about `DCHECK`s. I insist on using DCHECKs for `task` and `thread` everywhere because according to `Timer` class design these two pointers **should** be alive during **all** `Timer`class life. So I don't found them as redundant

**Comment by AByzhynar** [comment]
Reviewed

**Comment by AGritsevich** [comment]
Reviewed

**Comment by pestOO** [comment]
@anosach-luxoft, please squash and split commits before merge

File: src/components/utils/src/threads/posix_thread.cc [hunk 22]
```diff
@@ -258,7 +262,7 @@ void Thread::stop() {
 
 void Thread::join() {
   LOG4CXX_AUTO_TRACE(logger_);
-  DCHECK(!pthread_equal(pthread_self(), handle_));
+  DCHECK(!IsCurrentThread());
```
Comment (pestOO) [inline src/components/utils/src/threads/posix_thread.cc]: @anosach-luxoft, May be use `DCHECK_OR_RETUNR`?
In release we will get an leaked thread in stead of deadlock.

**Comment by anosach-luxoft** [comment]
@pestOO, regarding your concern about `Thread` class. This point will be investigated within appropriate [issue](https://adc.luxoft.com/jira/browse/APPLINK-23297) with my own concerns.

File: src/components/utils/src/timer.cc (hunk 12 above)
Comment (pestOO) [inline src/components/utils/src/timer.cc]: It looks like we need to delete DCHECK at all all and use returns version only

**Comment by anosach-luxoft** [comment]
Guys, please remind me if I've forgot some of your comments

**Comment by pestOO** [comment]
@anosach-luxoft, thanks for ticket!

**Comment by pestOO** [comment]
@anosach-luxoft, please review _all_ comment - at least few about DCHECK are not answered.

**Comment by anosach-luxoft** [comment]
@pestOO 
Few words about DCHECKs. I insist on using DCHECKs for task and thread everywhere because according to Timer class design these two pointers should be alive during all Timerclass life. So I don't found them as redundant

**Comment by anosach-luxoft** [comment]
@pestOO please see 5007765. After your confirmation I'll squash commits and merge

**Comment by pestOO** [comment]
Reviewed
